### PR TITLE
Pybind 11: count_subscribers and count_publishers

### PR DESCRIPTION
**Pull request by ahcorde**
Part of #665

Handles:

```
rclpy_count_subscribers
rclpy_count_publishers
```

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
Building up-to `rclpy` and testing `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14043/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8864/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11736/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14134/)

File: rclpy/src/rclpy/node.cpp [hunk 1]
```diff
@@ -62,4 +63,43 @@ get_node_namespace(py::capsule pynode)
 
   return node_namespace;
 }
+
+unsigned int
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:67]: @ahcorde ni: doesn't `size_t` work?

File: rclpy/src/rclpy/node.cpp [hunk 2]
```diff
@@ -62,4 +63,43 @@ get_node_namespace(py::capsule pynode)
...
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  size_t count = 0;
+  rcl_ret_t ret = rcl_count_publishers(node, topic_name, &count);
+  if (ret != RCL_RET_OK) {
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:78]: @ahcorde nit: flip operands as per style guide. Same elsewhere.

File: rclpy/src/rclpy/node.cpp [hunk 3]
```diff
@@ -62,4 +63,43 @@ get_node_namespace(py::capsule pynode)
...
+    throw RCLError(std::string("Error in rcl_count_publishers: ") +
+      rcl_get_error_string().str);
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:80]: @ahcorde nit: `RCLError()` already collects the `rcl` error string. Same elsewhere.

File: rclpy/src/rclpy/node.hpp [hunk 4]
```diff
@@ -42,6 +42,33 @@ get_node_name(py::capsule pynode);
  */
 const char *
 get_node_namespace(py::capsule pynode);
+
+/// Count publishers for a topic.
+/*
+ * Raises ValueError if pynode is not a node capsule
+ * Raises RCLError if rcl_count_publishers doesn't work
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:49]: @ahcorde :smile:, would you consider: 
```suggestion
 * Raises RCLError if an error occurs in rcl
```
? Same below.

File: rclpy/src/rclpy/node.hpp [hunk 5]
```diff
@@ -42,6 +42,33 @@ get_node_name(py::capsule pynode);
...
+ *
+ * \param[in] pynode Capsule pointing to the node to get number of subscribers
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:51]: @ahcorde nit: 
```suggestion
 * \param[in] pynode Capsule pointing to a node
```

File: rclpy/src/rclpy/node.hpp [hunk 6]
```diff
@@ -42,6 +42,33 @@ get_node_name(py::capsule pynode);
...
+ */
+unsigned int
+get_count_publishers(py::capsule pynode, const char * topic_name);
+
+/// Count subscribers for a topic.
+/*
+ * Raises ValueError if pynode is not a node capsule
+ * Raises RCLError if rcl_count_subscribers doesn't work
+ *
+ * \param[in] pynode Capsule pointing to the node to get number of subscribers
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:64]: @ahcorde nit: 
```suggestion
 * \param[in] pynode Capsule pointing to a node
```

File: rclpy/src/rclpy/node.hpp [hunk 7]
```diff
@@ -42,6 +42,33 @@ get_node_name(py::capsule pynode);
  */
 const char *
 get_node_namespace(py::capsule pynode);
+
+/// Count publishers for a topic.
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:46]: @ahcorde meta: I think it'd be worth mentioning that this function returns the count of all the publishers known for that topic in the entire ROS graph.

File: rclpy/src/rclpy/node.hpp [hunk 8]
```diff
@@ -42,6 +42,33 @@ get_node_name(py::capsule pynode);
...
+ *
+ * \param[in] pynode Capsule pointing to the node to get number of subscribers
+ * for a topic
+ * \param[in] topic_name Name of the topic to count the number of publishers
+ * \return count of publishers
+ */
+unsigned int
+get_count_publishers(py::capsule pynode, const char * topic_name);
+
+/// Count subscribers for a topic.
```
Comment (hidmic) [inline rclpy/src/rclpy/node.hpp:59]: @ahcorde meta: I think it'd be worth to mention that this function returns the count of all the subscribers known for that topic in the entire ROS graph.

File: rclpy/src/rclpy/node.hpp (hunk 7 above)
Comment (hidmic) [reply rclpy/src/rclpy/node.hpp:46]: And replicate that extra information in the docstring.

File: rclpy/src/rclpy/node.hpp (hunk 8 above)
Comment (hidmic) [reply rclpy/src/rclpy/node.hpp:59]: And replicate that extra information in the docstring.

**Comment by hidmic** [comment]
@ahcorde `uncrustify` doesn't seem to be happy.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14046/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8867/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11739/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14137/)

**Review by hidmic — APPROVED** [review]
LGTM !

**Review by sloretz — APPROVED** [review]

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14061/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8881/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11753/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14152/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14065/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8885/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11757/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14157/)

**Pull request merged**